File: 0383-ransom-note/0383-ransom-note.dart
class Solution {
  bool canConstruct(String ransomNote, String magazine) {

    Map<String, int> ransomNoteMap = {};
    Map<String, int> magazineMap = {};

    for (int i = 0; i < ransomNote.length; ++i) {
      ransomNoteMap[ransomNote[i]] = (ransomNoteMap[ransomNote[i]] ?? 0) + 1;
    }
    for (int i = 0; i < magazine.length; ++i) {
      magazineMap[magazine[i]] = (magazineMap[magazine[i]] ?? 0) + 1;
    }

    for (String char in ransomNoteMap.keys) {
      if ((magazineMap[char] ?? 0) < ransomNoteMap[char]!) {
        return false;
      }
    }

    return true;
  }
}

File: 0383-ransom-note/0383-ransom-note.swift
class Solution {
    func canConstruct(_ ransomNote: String, _ magazine: String) -> Bool {
        var freqNote: [Character : Int] = [:]
        var freqMaga: [Character : Int] = [:]
        for character in ransomNote {
            freqNote[character] = freqNote[character, default:0] + 1;
        } 
        for character in magazine {
            freqMaga[character] = freqMaga[character, default:0] + 1;
        } 
        for character in ransomNote {
            if let magazineCount = freqMaga[character], let noteCount = freqNote[character] {
                if magazineCount >= noteCount {
                    continue 
            } else {
                return false
            }
        } else {
            return false
        }
}
        return true
    }
}